libyul/optimiser/KnowledgeBase.cpp

```diff
@@ -80,29 +71,130 @@ bool KnowledgeBase::knownToBeZero(YulString _a)
+		// if we already computed it.
+		if (VariableOffset const* varOff = util::valueOrNullptr(m_offsets, _var))
+			return *varOff;
+		value = valueOf(_var);
+	}
+	else
+	{
+		// For non-SSA, we query the value first so that the variable is reset if it has changed
+		// since the last call.
+		value = valueOf(_var);
+		if (VariableOffset const* varOff = util::valueOrNullptr(m_offsets, _var))
+			return *varOff;
+	}
+
+	if (value)
+		if (optional<VariableOffset> offset = explore(*value))
+			return setOffset(_var, *offset);
+	return setOffset(_var, VariableOffset{_var, 0});
+
 }
 
-Expression KnowledgeBase::simplifyRecursively(Expression _expression)
+optional<KnowledgeBase::VariableOffset> KnowledgeBase::explore(Expression const& _value)
 {
-	if (m_counter++ > 100)
-		return _expression;
+	if (Literal const* literal = std::get_if<Literal>(&_value))
+		return VariableOffset{YulString{}, valueOfLiteral(*literal)};
+	else if (Identifier const* identifier = std::get_if<Identifier>(&_value))
+		return explore(identifier->name);
+	else if (FunctionCall const* f = get_if<FunctionCall>(&_value))
+		if (f->functionName.name == "add"_yulstring || f->functionName.name == "sub"_yulstring)
+			if (optional<VariableOffset> a = explore(f->arguments[0]))
+				if (optional<VariableOffset> b = explore(f->arguments[1]))
+				{
+					u256 offset =
+						f->functionName.name == "add"_yulstring ?
+						a->offset + b->offset :
+						a->offset - b->offset;
+					if (a->reference == b->reference)
+						// Offsets relative to the same reference variable
+						return VariableOffset{a->reference, offset};
+					else if (a->reference == YulString{})
+						// a is constant
+						return VariableOffset{b->reference, offset};
+					else if (b->reference == YulString{})
+						// b is constant
+						return VariableOffset{a->reference, offset};
+				}
 
-	if (holds_alternative<FunctionCall>(_expression))
-		for (Expression& arg: std::get<FunctionCall>(_expression).arguments)
-			arg = simplifyRecursively(arg);
+	return {};
+}
 
-	if (auto match = SimplificationRules::findFirstMatch(_expression, m_dialect, m_variableValues))
-		return simplifyRecursively(match->action().toExpression(debugDataOf(_expression)));
+Expression const* KnowledgeBase::valueOf(YulString _var)
+{
+	AssignedValue const* assignedValue = m_variableValues(_var);
+	Expression const* currentValue = assignedValue ? assignedValue->value : nullptr;
+	if (m_valuesAreSSA)
+		return currentValue;
+
+	Expression const* lastValue = m_lastKnownValue[_var];
+	if (lastValue != currentValue)
+		reset(_var);
+	m_lastKnownValue[_var] = currentValue;
+	return currentValue;
+}
 
-	return _expression;
+void KnowledgeBase::reset(YulString _var)
+{
+	yulAssert(!m_valuesAreSSA);
+
+	m_lastKnownValue.erase(_var);
+	if (VariableOffset const* offset = util::valueOrNullptr(m_offsets, _var))
+	{
+		// Remove var from its group
+		if (offset->reference != YulString{})
```

Comment:
I kinda don't like that you're default constructing a `YulString` instance just make a check. Would it make more sense have a static instance such as `YulString::Default()` and compare against that?

Reply:
I can also use `.empty()`

Reply:
Huh, missed that. Even better. I counted 3 or 4 instances like this by the way.